### PR TITLE
Create yum repository files when building packages

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: lib/mockbuilder.py [hunk 1]
```diff
@@ -195,12 +203,32 @@ def copy_results(self, package):
...
+    def create_repository(self):
+        """
+        Create yum repository in build results directory.
+        """
+        result_dir = CONF.get('common').get('result_dir')
+        build_results_dir = os.path.join(
+            result_dir, 'packages', self.timestamp)
+        utils.run_command("createrepo %s" % build_results_dir)
+
+        repo_files_dir = os.path.join(result_dir, "repository_files")
```
Comment (mopsfelder) [inline lib/mockbuilder.py]: I don't like bike-shedding discussion but `repository_files` sounds awkward to me.

What if we shortened it to just `repository` then `result` directory would look like the following:

```
result/repository/
result/iso/
...
```

It's implicit that some *files* will be inside *repository* directory, we don't need to re-state that in the directory name.

We can still change it a little bit to something similar to what is done in [CentOS image](http://mirror.centos.org/altarch/7/os/ppc64le/):

```
result/packages/
result/repodata/
result/iso/
...
```

What do you think?

Comment (ghost) [reply lib/mockbuilder.py]: The actual repository is saved to `result/packages/<timestamp>`
The `result/repository_files` directory is supposed to keep only the `.repo` files, to ease the utilization of the repositories.
Another option would be to add another level to the former directory, i.e. `result/packages/<timestamp>/repository`, and putting the `.repo` file in `result/packages/<timestamp>`. But I prefer to keep them separated.

File: lib/mockbuilder.py [hunk 2]
```diff
@@ -199,16 +207,28 @@ def create_repository(self):
         """
         Create yum repository in build results directory.
         """
+        result_dir = CONF.get('common').get('result_dir')
         build_results_dir = os.path.join(
-            CONF.get('common').get('result_dir'), 'packages', self.timestamp)
+            result_dir, 'packages', self.timestamp)
         utils.run_command("createrepo %s" % build_results_dir)
 
+        repo_files_dir = os.path.join(result_dir, "repository_files")
```
Comment (alanoe) [inline lib/mockbuilder.py]: - Why "files" in "repository_files" if there is only one file?
- Agree with Murilo that "files" do not add anything
- It seems you are assuming that "repository_files" == .repo file and "repository" == packages + repository metadata files automatically created by repodata. 

Could we make this clearer? I suggest repository_files => repository_config

File: lib/mockbuilder.py [hunk 3]
```diff
@@ -195,6 +195,14 @@ def copy_results(self, package):
             self.timestamp, package.name)
         self._copy_rpms(package.build_cache_dir, package_build_results_dir)
 
+    def create_repository(self):
+        """
+        Create yum repository in build results directory.
+        """
+        build_results_dir = os.path.join(
+            CONF.get('common').get('result_dir'), 'packages', self.timestamp)
+        utils.run_command("createrepo %s" % build_results_dir)
```
Comment (alanoe) [inline lib/mockbuilder.py]: minor: YUM repository creation has nothing to do with Mock, IMHO this should be in a different Python module.

File: lib/mockbuilder.py [hunk 4]
```diff
@@ -30,6 +30,14 @@
 CONF = config.get_config().CONF
```
Comment (alanoe) [inline lib/mockbuilder.py:30]: Regarding commit msg, it took me a while to understand that "yum repository files" here do not refer to the yum repository metadata files, but only to the yum repository config file. I'd prefer "Create yum repository config file when building packages"

File: lib/mockbuilder.py (hunk 1 above)
Comment (mopsfelder) [reply lib/mockbuilder.py]: Perhaps the timestamp should be the father of the resulting descending directories of a build, for example:

```
result/<timestamp>/packages/<file>.rpm
result/<timestamp>/repository/<file>.repo
result/<timestamp>/repository/repodata/<file>.xml
result/<timestamp>/iso/<file>.iso
```

This facilitates if we want to move the output of a build to somewhere, i.e. we would just need to move the `result/<timestamp>` directory and all output will be together, no need to pickup sparse directories in the filesystem.

File: README.rst [hunk 5]
```diff
@@ -128,60 +128,31 @@ file or disable package repositories update in config.yaml.
 Using the RPMs
 --------------
 
-We decided that's is a little bit intrusive to install all the
-produced RPMs since some of them may not fit your needs. At the end of
```
Comment (alanoe) [inline README.rst]: Should we remove this first sentence?

File: README.rst [hunk 6]
```diff
@@ -128,60 +128,31 @@ file or disable package repositories update in config.yaml.
...
-the process you'll see a new ``result`` directory inside this
-project's root.
+At the end of the process you'll see a new ``result`` directory inside
+this project's root. It contains the built packages repositories under
+``packages`` and yum repository files under ``repository_files`` to
```
Comment (alanoe) [inline README.rst]: yum repository files => a yum repository config file

to avoid ambiguity

Also, repository_files => repository_config

File: README.rst [hunk 7]
```diff
@@ -128,60 +128,31 @@ file or disable package repositories update in config.yaml.
...
+simplify using those repositories.
 
-A suggested set of packages tested is the following:
+To install only virtualization related packages, for example, you can
```
Comment (alanoe) [inline README.rst]: - You are altering 2 things in the same patch: mentioning the new yum repository config file and mentioning the new metapackages. IMHO the latter should be in a separate patch.
- It'd be nice to mention the list of RPM packages installed by each metapackage in some doc (AFAIK it is not) and not force the user to read rpm .spec files to figure it out.

**Review by alanoe — CHANGES_REQUESTED** [review]

File: lib/mockbuilder.py (hunk 1 above)
Comment (alanoe) [reply lib/mockbuilder.py]: I like this ^ . It deviates from traditional yum repo organization, in which the repodata dir is in the same dir of the packages, but I dont know if that is a problem.

File: lib/mockbuilder.py (hunk 4 above)
Comment (ghost) [reply lib/mockbuilder.py:30]: Done.

File: lib/mockbuilder.py (hunk 2 above)
Comment (ghost) [reply lib/mockbuilder.py]: Renamed to repository_config

File: README.rst (hunk 6 above)
Comment (ghost) [reply README.rst]: Done.

File: README.rst (hunk 7 above)
Comment (ghost) [reply README.rst]: - Separated them.
- I agree. Those lists should be in the versions repository and it would be nice if they were updated automatically with the `update-versions` command.

File: lib/mockbuilder.py (hunk 3 above)
Comment (ghost) [reply lib/mockbuilder.py]: I don't currently see the benefit of creating a module for a single function that will only be called from here.

File: README.rst (hunk 5 above)
Comment (ghost) [reply README.rst]: We are recommending only the virtualization packages and the user is executing the install command himself, so he has control over what's installed. I think this sentence is unnecessary.

**Comment by olavphibm** [comment]
start build

File: lib/mockbuilder.py (hunk 3 above)
Comment (alanoe) [reply lib/mockbuilder.py]: IMHO it makes more sense to put this in utils.py or a new yum_repository.py. You could put the yum repository config file creation there too in the later patch which does it. This module should contain *only* Mock-specific features. But up to you.

**Review by alanoe — APPROVED** [review]

**Review by lisibann — APPROVED** [review]

File: lib/mockbuilder.py (hunk 1 above)
Comment (ghost) [reply lib/mockbuilder.py]: @mopsfelder I don't know if that's a good idea. With this structure, we wouldn't know if `result/latest` points to a `builds-packages` result or a `build-iso` result. The `repository_config` files that I'm creating here are also only useful in the system in which the build was executed. It would make no sense to move it around since the paths have to be hardcoded in yum config.

**Review by farosas — APPROVED** [review]
